MenuItem.cs

```diff
@@ -247,66 +255,103 @@ private static void ClickHandler(object sender, EventArgs e)
+            if (Settings.DisplayNicknames || Settings.DisplayCustomNicknames)
+            {
+                // Hide group's nickname for now
+
+                if (obj is GH_Panel || obj is GH_Group || obj is GH_NumberSlider)
+                    return false;
+
+                // TODO: Draw Group's nickname at the largest font size
+                // in the group's container box to be able to read it when the zoom is too low.
+            }
+
+            // Move filter logic ahead to enable filter on custom nicknames.
+
+            // Do not split the string in every paint call. Cache them in Settings.
+            // String manipulation & IEnumerable.Contains are time-costly.
+            if (Settings.IsFilterCustomEnabled &&
+                Settings.ShouldExcludeObject(obj.Name))
+                return false;
+
+            if (!Settings.FilterComponents &&
+                (obj is GH_Cluster || obj is IGH_Component || obj.ComponentGuid == galapagosID))
+                return false;
+
+            if (!Settings.FilterSpecial &&
+                IsSpecialObject(obj))
+                return false;
+
+            if (!Settings.FilterParameters &&
+                obj is IGH_Param)
+                return false;
+
+            if (!Settings.FilterGraphic &&
+                !(obj is IGH_ActiveObject))
+                return false;
+
             if (Settings.DisplayCustomNicknames)
             {
                 var code = string.Join(".", obj.Category, obj.SubCategory, obj.Name);
-                var defNick = string.Empty; 
-                if (_nicknamesCache.ContainsKey(code))
-                {
-                    defNick = _nicknamesCache[code];
-                }
-                else
+
+                // Reduce string manipulation to improve performance
+                var id = obj.ComponentGuid;
+
+                if (!_nicknamesCache.TryGetValue(code, out var defNick))
                 {
-                    var id = obj.ComponentGuid;
                     IGH_ObjectProxy proxy = null;
-                    if(id == clusterID || id == vbID || id == csID || id == pyID || id == Guid.Empty)
+                    if (id == clusterID || id == vbID || id == csID || id == pyID || id == Guid.Empty)
                     {
                         proxy = Grasshopper.Instances.ComponentServer.FindObjectByName(obj.Name, true, true);
                     }
                     else
                     {
                         proxy = Grasshopper.Instances.ComponentServer.EmitObjectProxy(id);
                     }
-                  
-                    if(proxy != null)
+
+                    if (proxy != null)
                     {
-                        defNick = proxy.Desc.NickName; 
+                        defNick = proxy.Desc.NickName;
                         _nicknamesCache.Add(code, defNick);
                     }
                     else
                     {
+                        if (obj is GH_Cluster)
```

Comment:
Deal with the special case when it's a newly-created cluster or user object not currently loaded.
My code assumes it is a newly-created one. I believe there's no way of knowing whether the nickname is changed if the cluster is a not-currently-loaded user object.

Reply:
New clusters must display its name/nickname except if are excluded with Filter Objects > Exclusions.